jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -259,26 +280,47 @@ private int getQueueSize()
         }
     }
 
-    @Override
-    public void succeeded()
-    {
-        succeedEntries();
-        super.succeeded();
-    }
-
-    private boolean succeedEntries()
+    public void timeoutExpired()
     {
-        boolean hadEntries = false;
-        for (Entry entry : entries)
+        boolean failed = false;
+        synchronized (FrameFlusher.this)
         {
-            hadEntries = true;
-            if (entry.frame.getOpCode() == OpCode.CLOSE)
-                endPoint.shutdownOutput();
-            notifyCallbackSuccess(entry.callback);
-            entry.release();
+            if (closedCause != null)
+                return;
+
+            long time = System.currentTimeMillis();
+            long nextTimeout = -1;
+
+            Iterator<Entry> iterator = TypeUtil.concat(entries.iterator(), queue.iterator());
+
+            while (iterator.hasNext())
+            {
+                Entry entry = iterator.next();
+                long timeSinceCreation = time - entry.getTimeOfCreation();
+
+                if (timeSinceCreation >= idleTimeout)
+                {
+                    LOG.warn("FrameFlusher write timeout on entry: {}", entry);
+                    failed = true;
+                    canEnqueue = false;
+                    closedCause = new WebSocketWriteTimeoutException("FrameFlusher Write Timeout");
+                    failedEntries.addAll(entries);
+                    failedEntries.addAll(queue);
+                    entries.clear();
+                    queue.clear();
+                    break;
+                }
+
+                if (timeSinceCreation > nextTimeout)
+                    nextTimeout = timeSinceCreation;
+            }
+
+            if (!failed && idleTimeout>0 && !entries.isEmpty())
+                timeoutScheduler.schedule(this::timeoutExpired, nextTimeout, TimeUnit.MILLISECONDS);
         }
-        entries.clear();
-        return hadEntries;
+
+        if (failed)
+            this.iterate();
```

Comment:
not sure this is really necessary

Reply:
if not, then who processes the failedEntries ?